modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/exec/rel/FullOuterJoinNode.java

```diff
@@ -0,0 +1,281 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.internal.processors.query.calcite.exec.rel;
+
+import java.util.ArrayDeque;
+import java.util.ArrayList;
+import java.util.Deque;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Set;
+import java.util.function.Predicate;
+import org.apache.ignite.internal.processors.query.calcite.exec.ExecutionContext;
+import org.apache.ignite.internal.processors.query.calcite.exec.RowHandler;
+import org.apache.ignite.internal.util.typedef.F;
+
+/** */
+public class FullOuterJoinNode<Row> extends AbstractNode<Row> {
+    /** Left row factory. */
+    private final RowHandler.RowFactory<Row> leftRowFactory;
+
+    /** Right row factory. */
+    private final RowHandler.RowFactory<Row> rightRowFactory;
+
+    /** Whether current left row was matched or not. */
+    private boolean leftMatched;
+
+    /** */
+    private final Set<Row> rightNotMatched = new HashSet<>();
```

Comment:
Using `HashSet` here looks like not a good idea because in general `Row` has not overridden hc/e methods. Especially current row implementation is a simple `Object[]`. I agree that we need a semantics of set here, but we need to think of its implementation more thoroughly.

Reply:
Actually it was what I relied on. Identity hash code and reference equality should be used here. Anyway I just rewrote this part with BitSet, so fixed